src/msix/msix.cpp

```diff
@@ -288,4 +289,52 @@ MSIX_API HRESULT STDMETHODCALLTYPE PackPackage(
     return static_cast<HRESULT>(MSIX::Error::OK);
 } CATCH_RETURN();
 
+MSIX_API HRESULT STDMETHODCALLTYPE SignPackage(
+    MSIX_SIGNING_OPTIONS signingOptions,
+    LPCSTR package,
+    MSIX_CERTIFICATE_FORMAT signingCertificateFormat,
+    LPCSTR signingCertificate,
+    LPCSTR privateKey
+) noexcept try
+{
+    ThrowErrorIf(MSIX::Error::InvalidParameter,
+        (package == nullptr || signingCertificate == nullptr),
+        "Invalid parameters");
+
+    if (signingCertificateFormat == MSIX_CERTIFICATE_FORMAT::MSIX_CERTIFICATE_FORMAT_UNKNOWN)
+    {
+        signingCertificateFormat = MSIX::DetermineCertificateFormat(signingCertificate);
+
+        ThrowErrorIf(MSIX::Error::InvalidParameter,
+            signingCertificateFormat == MSIX_CERTIFICATE_FORMAT::MSIX_CERTIFICATE_FORMAT_UNKNOWN,
+            "Certificate format could not be determined");
+
+        ThrowErrorIf(MSIX::Error::InvalidParameter,
+            (MSIX::DoesCertificateFormatRequirePrivateKey(signingCertificateFormat) && privateKey == nullptr),
+            "Certificate format requires separate private key");
+    }
+
+    MSIX::ComPtr<IStream> packageStream = 
+        MSIX::ComPtr<IStream>::Make<MSIX::FileStream>(MSIX::utf8_to_wstring(package).c_str(), MSIX::FileStream::Mode::READ_UPDATE);

```

Comment:
>utf8_to_wstring [](start = 60, length = 15)

do we need to use the wstring constructor of FileStream?

Reply:
Just following the pattern of CreateStreamOnFile

---
In reply to: [315937292](https://github.com/Microsoft/msix-packaging/pull/195#discussion_r315937292) [](ancestors = 315937292)